capsules/extra/src/tickv.rs

```diff
@@ -39,12 +33,165 @@
 use core::cell::Cell;
 use kernel::hil::flash::{self, Flash};
 use kernel::hil::hasher::{self, Hasher};
-use kernel::hil::kv_system::{self, KVSystem};
 use kernel::utilities::cells::{MapCell, OptionalCell, TakeCell};
 use kernel::utilities::leasable_buffer::{SubSlice, SubSliceMut};
 use kernel::ErrorCode;
 use tickv::{self, AsyncTicKV};
 
+/// The type of keys, this should define the output size of the digest
+/// operations.
+pub trait KeyType: Eq + Copy + Clone + Sized + AsRef<[u8]> + AsMut<[u8]> {}
+
+impl KeyType for [u8; 8] {}
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait KVSystemClient<K: KeyType> {
```

Comment:
I don't really follow why this is moved from the HIL. Won't other backend implementations want something similar?

Reply:
@alistair23 I think this PR is basically redefining which interface it makes sense for alternative backends to target.

A client of KVSystem (in this PR `tickv_kv_store` spends much of it's effort basically translating keys between high-level keys (byte strings) and TickV-keys using the asynchronous `generate_key` API which, I think, is very specific to TicKV. I suspect alternative back-ends are unlikely to use a similar mechanism for indexing the datastore, so this interface just wouldn't make sense for them.

Of course, if that proves wrong, and `tickv_kv_store.rs` would otherwise be just copy-pasted for a different KV store we can (a) revisit this or (b) move the trait somewhere more shared.